Implement Gibbs Sampling for a multidim gaussian generative joint, by using the conditionals which are also gaussian distributions . The minimum requirement is for joint to have D=2 variables and for Gibbs to alternate between the two.

Define the parameters of the joint Gaussian distribution

In [28]:
import numpy as np

mu = np.array([4.0, 5.0])
std = np.array([1.0, 2.0])

coef = 0.6

Precompute conditional distribution parameters

In [29]:

cond_var_x = std[0]**2 * (1 - coef**2)  # Variance of X|Y
cond_var_y = std[1]**2 * (1 - coef**2)  # Variance of Y|X


Initialize the sampler

In [30]:
x, y = 0, 0
n_samples = 10000
burn_in = 100
samples = np.zeros((n_samples, 2))

Run Gibbs sampling

In [ ]:

for i in range(n_samples + burn_in):
    conditional_mu_x = mu[0] + coef * (std[0]/std[1]) * (y - mu[1])
    x = np.random.normal(conditional_mu_x, np.sqrt(cond_var_x))
    
    conditional_mu_y = mu[1] + coef * (std[1]/std[0]) * (x - mu[0])
    y = np.random.normal(conditional_mu_y, np.sqrt(cond_var_y))
    

    if i >= burn_in:
        samples[i - burn_in] = [x, y]

print("Sample mean:", np.mean(samples, axis=0))
print("Sample std:", np.std(samples, axis=0))
print("Sample covariance:\n", np.cov(samples.T))

Sample mean: [4.00804089 5.00785336]
Sample std: [0.99950319 2.01457585]
Sample covariance:
 [[0.99910653 1.21629491]
 [1.21629491 4.05892176]]


Mu : 4.01, 5.02 -> Joint Probability Mu : 4, 5 (o)
std : 0.99, 2.01 -> Joint Probability std : 1, 2 (o)